### PR TITLE
Transfer learning with pytorch engine on fresh fruit dataset

### PR DESCRIPTION
**Pull request by KexinFeng**
## Description ##
In this PR, the following features are enabled:
1. The transfer learning feature on PyTorch engine. User can opt in this feature by setting "retrain" to String "true" in `Criteria.builder().optOption(String key, String value)`, when loading the traced pretrained model.
2. The trainable variables(parameters) can have per variable loss function.

### CODE REVIEW

File: api/src/main/java/ai/djl/modality/cv/transform/ToOneHot.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ */
+package ai.djl.modality.cv.transform;
+
+import ai.djl.modality.cv.util.NDImageUtils;
+import ai.djl.ndarray.NDArray;
+import ai.djl.translate.Transform;
+
+/** A {@link Transform} that converts the labels {@link NDArray} to one-hot labels. */
+public class ToOneHot implements Transform {
+    private final int numClass;
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/transform/ToOneHot.java:21]:
```suggestion
    private int numClass;
```

File: api/src/main/java/ai/djl/modality/cv/transform/ToOneHot.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * and limitations under the License.
+ */
+package ai.djl.modality.cv.transform;
+
+import ai.djl.modality.cv.util.NDImageUtils;
+import ai.djl.ndarray.NDArray;
+import ai.djl.translate.Transform;
+
+/** A {@link Transform} that converts the labels {@link NDArray} to one-hot labels. */
+public class ToOneHot implements Transform {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/transform/ToOneHot.java:20]: I think it's better just call this class `OneHot`

File: api/src/main/java/ai/djl/modality/cv/util/NDImageUtils.java [hunk 3]
```diff
@@ -130,6 +130,19 @@ public static NDArray toTensor(NDArray image) {
...
+    /**
+     * Converts the labels {@link NDArray} to one-hot labels.
+     *
+     * <p>Converts the labels {@link NDArray} to one-hot labels.
+     *
+     * @param label the label to convert
+     * @param numClass the number of classes
+     * @return the converted label
+     */
+    public static NDArray toOneHot(NDArray label, int numClass) {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/util/NDImageUtils.java:142]: Why add to NDImageUtils?

Can you use `NDArray.oneHot()`

File: api/src/main/java/ai/djl/repository/zoo/Criteria.java [hunk 4]
```diff
@@ -627,7 +627,7 @@ public Builder<I, O> optArgument(String key, Object value) {
          * @param options the model loading options
          * @return this {@code Builder}
          */
-        public Builder<I, O> optOptions(Map<String, String> options) {
+        public Builder<I, O> optOption(Map<String, String> options) {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/repository/zoo/Criteria.java:630]: This should be `optOptions()`, this passing a map of options,
`optOption()` set one option at a time

File: api/src/main/java/ai/djl/training/loss/SoftmaxCrossEntropyLoss.java [hunk 5]
```diff
@@ -64,6 +67,22 @@ public SoftmaxCrossEntropyLoss(
         this.classAxis = classAxis;
         this.sparseLabel = sparseLabel;
         this.fromLogit = fromLogit;
+        this.fromSoftmax = false;
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/loss/SoftmaxCrossEntropyLoss.java:70]: no need to set to false explicitly

File: api/src/main/java/ai/djl/training/loss/SoftmaxCrossEntropyLoss.java [hunk 6]
```diff
@@ -31,6 +31,7 @@ public class SoftmaxCrossEntropyLoss extends Loss {
     private int classAxis;
     private boolean sparseLabel;
     private boolean fromLogit;
+    private boolean fromSoftmax;
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/loss/SoftmaxCrossEntropyLoss.java:34]: Does fromLogit and fromSoftmax mutually exclusion?

File: api/src/main/java/ai/djl/training/tracker/FixedPerVarTracker.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/tracker/FixedPerVarTracker.java:2]:
```suggestion
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: examples/build.gradle [hunk 8]
```diff
@@ -8,6 +8,7 @@ dependencies {
     implementation project(":basicdataset")
     implementation project(":model-zoo")
     implementation project(":extensions:timeseries")
+    implementation 'org.testng:testng:7.1.0'
```
Comment (frankfliu) [inline examples/build.gradle:11]: 1. Do not use 7.1.0, it has bug
2. testng is already included in testImplemenation
3. testng should not be included as implemention

File: examples/src/test/java/ai/djl/examples/training/TransferFreshFruitTest.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline examples/src/test/java/ai/djl/examples/training/TransferFreshFruitTest.java:2]:
```suggestion
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/main/java/ai/djl/repository/zoo/Criteria.java (hunk 4 above)
Comment (KexinFeng) [reply api/src/main/java/ai/djl/repository/zoo/Criteria.java:630]: ok

File: api/src/main/java/ai/djl/ndarray/internal/NDFormat.java [hunk 10]
```diff
@@ -60,8 +60,9 @@ public static String format(
         if (array.hasGradient()) {
             sb.append(" hasGradient");
         }
+
         if (DEBUG) {
-            return sb.toString();
+            sb.append(" ~>");
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/ndarray/internal/NDFormat.java:65]: We need return here. IntelliJ trys to evaluate ndarray in debugging mode, which makes IDE super slow.

File: api/src/main/java/ai/djl/training/LocalParameterServer.java [hunk 11]
```diff
@@ -48,17 +48,17 @@ public void update(String parameterId, NDArray[] grads, NDArray[] params) {
...
-        try (NDArray aggregatedGrad = grads[0].duplicate()) {
-            for (NDArray param : params) {
-                if (param.getDevice().equals(firstDevice)) {
-                    optimizer.update(parameterId, param, aggregatedGrad);
-                } else {
-                    try (NDArray gradSumCopy = aggregatedGrad.toDevice(param.getDevice(), true)) {
-                        optimizer.update(parameterId, param, gradSumCopy);
-                    }
-                }
+        NDArray aggregatedGrad = grads[0].duplicate();
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/LocalParameterServer.java:51]: Why not use try-with-resource?

File: api/src/main/java/ai/djl/training/LocalParameterServer.java [hunk 12]
```diff
@@ -48,17 +48,17 @@ public void update(String parameterId, NDArray[] grads, NDArray[] params) {
...
+        for (NDArray param : params) {
+            if (param.getDevice().equals(firstDevice)) {
+                optimizer.update(parameterId, param, aggregatedGrad);
+            } else {
+                NDArray gradSumCopy = aggregatedGrad.toDevice(param.getDevice(), true);
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/LocalParameterServer.java:56]: why?

File: examples/src/main/java/ai/djl/examples/training/transferlearning/TransferFreshFruit.java [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+    public static TrainingResult runExample(String[] args)
+            throws IOException, TranslateException, ModelException, URISyntaxException {
+        Arguments arguments = new Arguments().parseArgs(args);
+        if (arguments == null) {
+            return null;
+        }
+
+        // Also available at
+        // "https://mlrepo.djl.ai/model/cv/image_classification/ai/djl/pytorch/resnet18_embedding/0.1/traced_resnet18_embedding.pt.gz";
+        String modelUrls = "/Users/fenkexin/Desktop/transferDJL/code/base_nw.pt";
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/training/transferlearning/TransferFreshFruit.java:72]: Why use local file path here?

File: examples/src/main/java/ai/djl/examples/training/transferlearning/TransferFreshFruit.java [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+        model.close();
+        embedding.close();
+        return null;
+    }
+
+    private static ImageFolder getData(String subfolderName, String fruit, int batchSize)
+            throws TranslateException, IOException {
+        // The dataset is from <a
+        // href="https://www.kaggle.com/datasets/sriramr/fruits-fresh-and-rotten-for-classification">https://www.kaggle.com/datasets/sriramr/fruits-fresh-and-rotten-for-classification</a>
+        String folderUrl = "/Users/fenkexin/Desktop/transferDJL/code/data/" + fruit;
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/training/transferlearning/TransferFreshFruit.java:136]: Can we add this dataset int our repo?
We should not hard code local path

File: examples/src/test/java/ai/djl/examples/training/TransferFreshFruitTest.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+public class TransferFreshFruitTest {
+
+    // TransferFreshFruit requires manually downloading the dataset from <a
+    // href="https://www.kaggle.com/datasets/sriramr/fruits-fresh-and-rotten-for-classification">https://www.kaggle.com/datasets/sriramr/fruits-fresh-and-rotten-for-classification</a> so can't be automatically tested
+    @Test(enabled = false)
```
Comment (frankfliu) [inline examples/src/test/java/ai/djl/examples/training/TransferFreshFruitTest.java:31]: We should create a fruit-unittest dataset, which only include a few images for testing. See pikachu-unittest dataset.

File: api/src/main/java/ai/djl/ndarray/internal/NDFormat.java (hunk 10 above)
Comment (KexinFeng) [reply api/src/main/java/ai/djl/ndarray/internal/NDFormat.java]: Would it help if we bound the string with shorter limit? Currently it is already bounded. I found that showing the content of an NDArray (even partially) is quite often needed when debugging a model during training. 
It looks like Pycharm IDE enables evaluating the ndarray(tensor) on pytorch engine, too.

File: examples/src/main/java/ai/djl/examples/training/transferlearning/TransferFreshFruit.java (hunk 13 above)
Comment (KexinFeng) [reply examples/src/main/java/ai/djl/examples/training/transferlearning/TransferFreshFruit.java:72]: There is still some problem in the s3. It can be downloaded mannually but the code cannot find it yet.

File: examples/src/main/java/ai/djl/examples/training/transferlearning/TransferFreshFruit.java (hunk 14 above)
Comment (KexinFeng) [reply examples/src/main/java/ai/djl/examples/training/transferlearning/TransferFreshFruit.java:136]: Should we put it in S3 or in our repo. The training dataset can be big.

File: examples/src/test/java/ai/djl/examples/training/TransferFreshFruitTest.java (hunk 15 above)
Comment (KexinFeng) [reply examples/src/test/java/ai/djl/examples/training/TransferFreshFruitTest.java:31]: ok

File: api/src/main/java/ai/djl/training/LocalParameterServer.java (hunk 12 above)
Comment (KexinFeng) [reply api/src/main/java/ai/djl/training/LocalParameterServer.java:56]: Forgot to restore it.

**Review by zachgk — APPROVED** [review]
I updated this with a new ParameterTracker abstraction which I think should make the situation more maintainable and easier to get "correct" in the future rather than forgetting to update the parameter id. With this change, I think it looks good to merge
